pkg/k8s/synced/resources_test.go

```diff
@@ -0,0 +1,133 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+//go:build !privileged_tests
+
+package synced
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// waitForCacheTest is a table test case for testing WaitForCacheSyncWithTimeout.
+// Each test case will similate waiting for sync of every key in resourceNamesToSyncDuration
+// with their respective timeout duration.
+type waitForCacheTest struct {
+	timeout time.Duration
+	// This maps resource names (ex. "core/v1::Pods") onto the duration the simulated
+	// cache sync will take to complete.
+	resourceNamesToSyncDuration map[string]time.Duration
+	// Maps resource names to a duration to wait after init upon which an "event" will
+	// be emitted. Used to test cases where events cause pushing out of timeout.
+	resourceNamesToEvent map[string]time.Duration
+	// Array which is passed to WaitForCacheSync... function.
+	// Only resources in this array should cause return of timeout error when doing test.
+	resourceNames []string
+	expectErr     error
+	// Used to test edge cases where controller doesn't actually invoke BlockWaitGroupToSyncResources.
+	dontStartBlockWaitGroupToSyncResources bool
+}
+
+func TestWaitForCacheSyncWithTimeout(t *testing.T) {
+	unit := func(d int) time.Duration { return syncedPollPeriod * time.Duration(d) }
+	assert := assert.New(t)
+	for msg, test := range map[string]waitForCacheTest{
+		"Should complete due to event causing timeout to be extended past initial timeout": {
+			timeout: unit(5),
+			resourceNamesToSyncDuration: map[string]time.Duration{
+				"foo": unit(7),
+				"bar": unit(7),
+			},
+			resourceNamesToEvent: map[string]time.Duration{
+				"foo": unit(4),
+			},
+			resourceNames: []string{"foo"},
+		},
+		"Should timeout due to watched resource exceeding timeout": {
+			timeout: unit(1),
+			resourceNamesToSyncDuration: map[string]time.Duration{
+				"foo": unit(3),
+			},
+			resourceNamesToEvent: map[string]time.Duration{},
+			resourceNames:        []string{"foo"},
+			expectErr:            fmt.Errorf("timed out after 100ms, never received event for resource \"foo\""),
+		},
+		"Any one timeout should cause error": {
+			timeout: unit(5),
+			resourceNamesToSyncDuration: map[string]time.Duration{
+				"foo": unit(7),
+				"bar": unit(7),
+			},
+			resourceNamesToEvent: map[string]time.Duration{
+				"foo": unit(4),
+			},
+			resourceNames: []string{"foo", "bar"},
+			expectErr:     fmt.Errorf("timed out after 500ms, never received event for resource \"bar\""),
```

Comment:
We are always expecting `bar` to be part of the error msg because it's the last resource to timeout? I don't understand how this unit test is not flaking because I'd expect Go to pick from a map randomly, which it does, so I'd expect this error msg to actually change on different runs.